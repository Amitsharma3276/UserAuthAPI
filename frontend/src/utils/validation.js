export const validate = (values) => {
  const digits_only = (string) =>
    [...string].every((ch) => "0123456789".includes(ch));

  const alphabet = (string) =>
    [...string].every((ch) =>
      "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".includes(ch)
    );

  let retError = "";

  //NAME VALIDATIONS
  if (values.name === null || values.name.length === 0) {
    retError = retError.concat("Name is required \n");
  }

  if (!alphabet(values.name)) {
    retError = retError.concat("Invalid Name \n");
  }

  //PASSWORD VALIDATIONS
  if (values.password === "" || values.password === null) {
  } else if (
    values.password != null &&
    (values.password.length > 15 || values.password.length < 8)
  ) {
    retError = retError.concat("Weak password  \n");
  }

  //PERMANENT ADDRESS VALIDATION
  if (values.p_line_one === null || values.p_line_one.length === 0) {
    retError = retError.concat("Permanent Address Line 1 is required \n");
  }

  if (values.p_city === null || values.p_city.length === 0) {
    retError = retError.concat("Permanent Address City is required \n");
  }

  if (values.p_state === null || values.p_state.length === 0) {
    retError = retError.concat("Permanent Address State is required \n");
  }

  if (
    values.p_pincode === null ||
    values.p_pincode.length !== 6 ||
    (values.p_pincode.length != null && !digits_only(values.p_pincode))
  ) {
    retError = retError.concat(
      "Enter Valid 6 digit pincode for permanent address \n"
    );
  }

  //CURRENT ADDRESS VALIDATION(IF ENTERED)
  if (values.c_line_one !== null && values.c_line_one.length !== 0) {
    if (values.c_city === null || values.c_city.length === 0) {
      retError = retError.concat("Current Address City is required \n");
    }
    if (values.c_state === null || values.c_state.length === 0) {
      retError = retError.concat("Current Address State is required \n");
    }
    if (
      values.c_pincode === null ||
      values.c_pincode.length !== 6 ||
      (values.c_pincode.length !== 0 && !digits_only(values.c_pincode))
    ) {
      retError = retError.concat(
        "Enter Valid 6 digit pincode for current address \n"
      );
    }
  }

  //EMERGENCY NUMBER VALIDATION
  if (
    values.emergency_number === null ||
    values.emergency_number.length !== 10 ||
    (values.emergency_number != null && !digits_only(values.emergency_number))
  ) {
    retError = retError.concat("Enter valid 10 digit contact number \n");
  }

  return retError;
};

// const alphaNumeric = (string) =>
// [...string].every((ch) =>
//   "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789".includes(
//     ch
//   )
// );
