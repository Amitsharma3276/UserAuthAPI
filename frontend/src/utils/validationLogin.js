export const validationLogin = (values) => {
  let retError = "";
  console.log(values);
  if (values.username === "" || values.username === null) {
    retError = retError.concat("Enter Username  \n");
  }
  if (values.password === "" || values.password === null) {
    retError = retError.concat("Enter Password  \n");
    console.log("Hello");
  } else if (
    values.password != null &&
    (values.password.length > 15 || values.password.length < 8)
  ) {
    retError = retError.concat("Enter correct password  \n");
  }
  console.log(retError);

  return retError;
};
