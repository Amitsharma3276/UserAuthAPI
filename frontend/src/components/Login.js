import React, { useState } from "react";
import FormHeader from "../common/FormHeader";
import Input from "../common/Input";
import { loginService } from "../services/loginService";
import { Error } from "../common/Error";
import { validationLogin } from "../utils/validationLogin";
import { redirectUser } from "../common/redirectUser";

const Login = () => {
  const [userInfo, setUserInfo] = useState({
    username: "",
    password: "",
    error: "",
    data: "",
    redirectToHome: false,
  });

  const { username, password, error, redirectToHome } = userInfo;

  const handleChange = (user) => (e) => {
    setUserInfo({ ...userInfo, error: false, [user]: e.target.value });
  };

  const clickLogin = (e) => {
    e.preventDefault();
    let errorMsg = validationLogin(userInfo);
    if (errorMsg) {
      setUserInfo({ ...userInfo, error: errorMsg, redirectToHome: false });
      return;
    }
    setUserInfo({ ...userInfo, error: false });
    let status;
    loginService(userInfo)
      .then((response) => {
        console.log(response.status);
        status = response.status;
      })
      .then((data) => {
        if (status === 200) {
          console.log("logged in " + status);
          setUserInfo({ ...userInfo, redirectToHome: true, data: data });
        } else {
          setUserInfo({
            ...userInfo,
            error: "Invalid Credentials",
            redirectToHome: false,
          });
        }
      });
  };

  const formHeader = {
    title: <h3>Welcome to Team AAAA's webapp</h3>,
    message: <p>Please Signup with your credentials</p>,
  };
  return (
    <div className="auth-inner">
      <form>
        <FormHeader title={formHeader.title} message={formHeader.message} />

        <br />

        {Error(userInfo)}
        {redirectUser(userInfo, setUserInfo)}

        <Input
          label="Username"
          onChange={handleChange("username")}
          type="text"
          className="form-control"
          placeholder="Enter username"
          value={username}
          isRequired="true"
        />
        <Input
          label="Password"
          onChange={handleChange("password")}
          type="password"
          className="form-control"
          placeholder="Enter password"
          value={password}
          isRequired="true"
        />

        <button
          onClick={clickLogin}
          type="submit"
          className="btn btn-primary btn-block"
        >
          Login
        </button>
        <p className="forgot-password text-right">
          New User? <a href="/signup">Signup</a>
        </p>
      </form>
    </div>
  );
};

export default Login;
