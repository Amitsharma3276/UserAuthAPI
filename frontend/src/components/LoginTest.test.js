import React from "react";
import { shallow } from "enzyme";
import Login from "./Login";
import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { act } from "react-dom/test-utils";
import { render } from "@testing-library/react";
import { unmountComponentAtNode } from "react-dom";
import * as loginUtil from "../utils/validationLogin";

configure({ adapter: new Adapter() });
let container = null;
beforeEach(() => {
  //render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

// jest.mock("../utils/validationLogin", () => {
//   return "";
// });

describe("<Login /> with no props", () => {
  const wrapper = shallow(<Login />);

  //SNAPSHOT
  it("should match the snapshot", () => {
    act(() => {
      render(wrapper, container);
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  //USERNAME
  it("should have an username input field", () => {
    act(() => {
      render(wrapper, container);
    });
    expect(wrapper.find('Input[type="text"]')).toHaveLength(1);
  });

  it("should have proper props for username field", () => {
    act(() => {
      render(wrapper, container);
    });
    expect(wrapper.find('Input[type="text"]').props()).toEqual({
      label: "Username",
      onChange: expect.any(Function),
      type: "text",
      className: "form-control",
      placeholder: "Enter username",
      value: expect.any(String),
      isRequired: "true",
    });
  });

  it("Username onChangeHandler test", () => {
    act(() => {
      render(wrapper, container);
    });
    // console.log('Check#'+wrapper.find('Input[type="password"]').props('onChange'));
    wrapper.find('Input[type="text"]').simulate("change", {
      target: { value: "Dummy" },
    });

    expect(wrapper.find('Input[type="text"]').prop("value")).toEqual("Dummy");
  });

  //PASSWORD
  it("should have a password input field", () => {
    act(() => {
      render(wrapper, container);
    });
    expect(wrapper.find('Input[type="password"]')).toHaveLength(1);
  });

  it("should have proper props for password field", () => {
    act(() => {
      render(wrapper, container);
    });
    expect(wrapper.find('Input[type="password"]').props()).toEqual({
      label: "Password",
      onChange: expect.any(Function),
      type: "password",
      className: "form-control",
      placeholder: "Enter password",
      value: expect.any(String),
      isRequired: "true",
    });
  });

  it("Password onChangeHandler test", () => {
    act(() => {
      render(wrapper, container);
    });
    // console.log('Check#'+wrapper.find('Input[type="password"]').props('onChange'));
    wrapper.find('Input[type="password"]').simulate("change", {
      target: { value: "Dummy@123" },
    });

    expect(wrapper.find('Input[type="password"]').prop("value")).toEqual(
      "Dummy@123"
    );
  });

  //LOGIN BUTTON
  it("render login component", () => {
    act(() => {
      render(wrapper, container);
    });
    const button = document.querySelector("button");
    expect(button.innerHTML).toBe("Login");
  });

  it("ClickLogin Button test with no errors", () => {
    act(() => {
      render(wrapper, container);
    });
    const clickEvent = { preventDefault: jest.fn() };
    //const spy = jest.spyOn(wrapper.instance(), "validationLogin");
    wrapper.find("button").simulate("click", clickEvent);
    expect(clickEvent.preventDefault).toBeCalledTimes(1);
  });

  it("Login Validation", () => {
    const fakeData = {
      username: "dummy",
      password: "Dummy@123",
      error: "",
      data: "",
      redirectToHome: false,
    };

    const errorResult = loginUtil.validationLogin(fakeData);
    console.log(errorResult);
    expect(errorResult).toBe("");

    const errorMessage = "Invalid Credentials";
    const mockValidationLogin = jest.spyOn(loginUtil, "validationLogin");
    mockValidationLogin.mockResolvedValue(errorMessage);
    return loginUtil.validationLogin(fakeData).then((response) => {
      //console.log(response);
      expect(response).toBe(errorMessage);
      expect(mockValidationLogin).toHaveBeenCalledTimes(1);
      mockValidationLogin.mockRestore();
    });
  });

  it("Login Validation with errors", () => {
    const fakeNewData = {
      username: "dummy",
      password: "",
      error: "",
      data: "",
      redirectToHome: false,
    };

    const validationMock = jest.spyOn(loginUtil, "validationLogin");
    const result = validationMock(fakeNewData);
    expect(validationMock).toHaveBeenCalledTimes(1);
    expect(result).toBe("Enter Password  \n");

    const erroMessage = "Enter Password";
    const mockValidationLogin = jest.spyOn(loginUtil, "validationLogin");
    mockValidationLogin.mockResolvedValue(erroMessage);
    return loginUtil.validationLogin(fakeNewData).then((response) => {
      expect(response).toBe(erroMessage);
      expect(mockValidationLogin).toHaveBeenCalledTimes(2);
      mockValidationLogin.mockRestore();
    });
  });

  // it("LoginService test without errors", () => {
  //   const fakeNewData = {
  //     username: "dummy",
  //     password: "Dummy@123",
  //     error: "",
  //     data: "",
  //     redirectToHome: false,
  //   };

  //   const LoginServiceMock = jest.spyOn(global, "fetch");
  //   const result = validationMock(fakeNewData);
  //   expect(validationMock).toHaveBeenCalledTimes(1);
  //   expect(result).toBe("Enter Password  \n");

  //   const erroMessage = "Enter Password";
  //   const mockValidationLogin = jest.spyOn(loginUtil, "validationLogin");
  //   mockValidationLogin.mockResolvedValue(erroMessage);
  //   return loginUtil.validationLogin(fakeNewData).then((response) => {
  //     expect(response).toBe(erroMessage);
  //     expect(mockValidationLogin).toHaveBeenCalledTimes(2);
  //     mockValidationLogin.mockRestore();
  //   });
  // });
});
