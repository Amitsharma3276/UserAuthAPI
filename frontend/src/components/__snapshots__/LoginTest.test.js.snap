// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> with no props should match the snapshot 1`] = `"<div class=\\"auth-inner\\"><form><div><center><h3>Welcome to Team AAAA&#x27;s webapp</h3><p>Please Signup with your credentials</p></center><hr/></div><br/><div class=\\"alert alert-warning display-linebreak\\" style=\\"display:none\\"><div class=\\"display-linebreak\\" style=\\"white-space:pre-line\\"></div></div><div class=\\"form-group\\"><label>Username*</label><input type=\\"text\\" class=\\"form-control\\" placeholder=\\"Enter username\\"/></div><div class=\\"form-group\\"><label>Password*</label><input type=\\"password\\" class=\\"form-control\\" placeholder=\\"Enter password\\"/></div><button type=\\"submit\\" class=\\"btn btn-primary btn-block\\">Login</button><p class=\\"forgot-password text-right\\">New User? <a href=\\"/signup\\">Signup</a></p></form></div>"`;
